usr/sbin/p11sak/p11sak.c

```diff
@@ -3007,12 +3007,15 @@ static CK_RV add_attribute(CK_ATTRIBUTE_TYPE type, const void *value,
 
     tmp[*num_attrs].type = type;
     tmp[*num_attrs].ulValueLen = value_len;
-    tmp[*num_attrs].pValue = malloc(value_len);
-    if (tmp[*num_attrs].pValue == NULL) {
-        warnx("Failed to allocate memory attribute to add to list");
-        return CKR_HOST_MEMORY;
+    tmp[*num_attrs].pValue = NULL;
```

Comment:
This fix is good, and we should definitely keep it.

However, I still don't understand how it can happen that `add_attribute()` is called with an zero `value_len` for CKA_ID. I tried several key-gens but non of them leads to a CKA_ID of zero length.

I even tried with `--id ""` , which would produce a zero length CKA_ID attribute value, but this segfaults in `parse_id()` already, because `BN_hex2bn()` does not allocate a bignum on a zero length input string, thus it segfaults later at `BN_bn2binpad()` because `b` is NULL.

I'll fix this with a separate PR shortly.

BTW: This also caused a zero length `calloc`, which would be problematic in AIX, I guess. 

Please elaborate how it has happend in your case that `add_attribute()` is called with  a zero length value.

The only way I found to cause `add_attribute()` to be called with a zero length CKA_ID value is for `extract-cert-pubkey` when the certificate has CKA_ID=<empty>, then the pubic key object will also be created with a zero length CKA_ID attribute.

Is it this where it happend in your case ?

Reply:
p11sak fixes are in https://github.com/opencryptoki/opencryptoki/pull/755